django/http/request.py

```diff
@@ -176,15 +177,29 @@ def build_absolute_uri(self, location=None):
             location = '//%s' % self.get_full_path()
         bits = urlsplit(location)
         if not (bits.scheme and bits.netloc):
-            current_uri = '{scheme}://{host}{path}'.format(scheme=self.scheme,
-                                                           host=self.get_host(),
-                                                           path=self.path)
-            # Join the constructed URL with the provided location, which will
-            # allow the provided ``location`` to apply query strings to the
-            # base path as well as override the host, if it begins with //
-            location = urljoin(current_uri, location)
+            # Handle the most common and trivial case trivially. If the location is
+            # absolute and no scheme or host (netloc) is provided we can shortcut
+            # urljoin as long as no path segments are '.' or '..'.
+            if (bits.path.startswith('/') and not bits.scheme and not bits.netloc and
+                    '/./' not in bits.path and '/../' not in bits.path):
+                # If location starts with '//' but has no netloc that means we
+                # should reuse the schema and netloc from the current request.
+                # Strip the double slashes and continue as if it wasn't specified.
+                if location.startswith('//'):
```

Comment:
I don't get this, if location is `//foo` then `bits.netloc` will be `foo`. The only way to get here is if location is prefixed by 4 (or more) forward slashes i.e. `////foo` or `/////foo` etc. which seems like uncommon input and should probably just be handled by `urljoin`

Reply:
`if (bits.path.startswith('/') and not bits.scheme and not bits.netloc and '/./' not in bits.path and '/../' not in bits.path):`

Could potentially be rewritten to:

```
path_bits = bits.path.split('/')
if (path_bits[0] == '' and not {'.', '..'}.intersection(path_bits)):
```

Reply:
If location starts with `//` and there is no netloc and no schema, that means the original location started with `///`, we can just strip that since it means "same scheme, same host"